File: backend/src/patrones_de_diseño/template/plantilla.rb
#!/bin/ruby

class Operacion
	# Define una operación en dos pasos
    # abstractos que cada subclase concretará
    def initialize
        metodos_abstractos = [
            "self.suboperacion_1",
            "self.suboperacion_2"
        ]

        for metodo in metodos_abstractos do
            if !defined? metodo.to_sym
            	raise TypeError, "Debes declarar los métodos " +
                    "abstractos de la clase padre Operacion " +
                    "en #{self.class.name}"
            end
        end
    end

    def ejecutar
        self.suboperacion_1
        self.suboperacion_2
    end
end


class OperacionConcreta < Operacion
    # Implementa las operaciones
    # para seguir los pasos del algoritmo
    # específicos para esta subclase */
    def suboperacion_1
        puts("Primer paso en la operación")
    end

    def suboperacion_2
        puts("Segundo paso en la operación")
    end
end

if __FILE__ == $0
    operacion_concreta = OperacionConcreta.new
    operacion_concreta.ejecutar()
end

File: backend/src/patrones_de_diseño/template/plantilla.js
"use strict"

class Operacion {
	// Define una operación en dos pasos
    // abstractos que cada subclase concretará
    constructor() {
        if (this.constructor === Operacion) {
            var msg = "La clase asbtracta Operacion " +
                "no puede ser instanciada directamente."
            throw new TypeError(msg);
        }

        var metodos_abstractos = [
            this.suboperacion_1,
            this.suboperacion_2
        ]

        for (var i=0; i < metodos_abstractos.length; i++){
            if (metodos_abstractos[i] === undefined) {
                var msg = "Debes declarar los métodos " +
                     "de la clase Padre Operacion en " +
                    this.constructor.name;
                throw new TypeError(msg);
            }
        }
    }

    ejecutar() {
        this.suboperacion_1();
        this.suboperacion_2();
    }
}

class OperacionConcreta extends Operacion {
    /* Implementa las operaciones
    para seguir los pasos del algoritmo
    específicos para esta subclase */
    suboperacion_1() {
        console.log("Primer paso en la operación");
    }

    suboperacion_2() {
        console.log("Segundo paso en la operación");
    }
}

if (require.main == module) {
    var operacion_concreta = new OperacionConcreta();
    operacion_concreta.ejecutar();
}
